proposals/NNNN-high-level-interface-modal-view-controllers.md

```diff
@@ -0,0 +1,160 @@
+
+- (void)present:(SDLApplication *)application; // private method called by the VC manager
+
+@end
+```
+
+#### SDLAlertController
+
+The alert controller is used to abstract the RPC `Alert`. The properties match the parameters of the RPC.
+
+```objc
+@interface SDLAlertController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLTTSChunk *> *initialPrompt;
+@property (nonatomic, copy) NSArray<SDLSoftButton *> *buttons;
+
+@property (nonatomic) BOOL shouldPlayTone;
+@property (nonatomic) BOOL progressIndicator;
+
+@end
+``` 
+
+#### SDLAudioInputController
+
+The audio input controller is used to abstract the RPC `PerformAudioPassThru`, `OnAudioPassThru` and `EndAudioPassThru`. The app developer should be able to provide a block which is called for audio data. Furthermore a method called `dismiss` should help to end the audio pass thru.
+
+```objc
+@interface SDLAudioInputController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLTTSChunk *> *initialPrompt;
+@property (nonatomic, copy) SDLSamplingRate samplingRate;
+@property (nonatomic, copy) SDLBitsPerSample bitsPerSample;
+@property (nonatomic, copy) SDLAudioType audioType;
+@property (nonatomic) BOOL muteMediaSource;
+
+@property (nonatomic) (^onAudioData)(NSData *audioData);
+
+- (void)finish; // sends an EndAudioPassThru in order to stop audio input
+
+@end
+```
+
+#### SDLScrollableMessageController
+
+Similar to `Alert` this view controller takes care of the RPC `ScrollableMessage`.
+
+```objc
+@interface SDLScrollableMessageController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLSoftButton *> *buttons;
+
+@end
+```
+
+#### SDLSliderController
+
+This controller matches very closely to the view [`UISlider`](https://developer.apple.com/documentation/uikit/uislider) but for consistency to SDL it makes more sense to treat it as a modal controller. In order to work with the RPC `Slider` it is important to provide the selected value back to the caller. As of UIKit this is not done by using a custom completion handler but simply provide the result in a property of the modal view controller. As per mobile API the property `Slider.sliderFooter` is an array used for two purposes. Either it's a footer (single item) or a list of names representing slider values (number of items must match `.numTicks`). The mobile API allows `.numTicks` to be between 2 and 26. The value range for the parameter `.position` can between 1 and 26 which means the range of values is not more than 26. With the high level abstraction the modal controller can manage the specified range allowing different min and max values. Examples are ranges with negative which are mapped internally to the available range. Future versions of the mobile API could increase the range which would be managed by the slider controller.
+
+```objc
+@interface SDLSliderController : SDLModalViewController
+
+@property (nonatomic, copy) NSString *title; // mapped with Slider.sliderHeader
```

Comment:
Missing a footer property

Reply:
I should have mentioned that the footer is covered by the modalVC.message property and ignored if valueLabels is used.

Reply:
I think that's pretty confusing for developers. See my comment above about splitting `message` into each subclass. I think that this is a good case where we're going to cause developer confusion by having `message` in the subclass. I would never have guessed that if you hadn't told me.